### ISSUE TITLE
CLI always requires a valid config for both swagger & routes

### ISSUE THREAD

**Issue by HoldYourWaffle** [lukeautry/tsoa#447]
I know the title might sound a little strange, but this really is an issue (albeit with an easy workaround).

While testing for #192 I struggled a bit with getting the configuration right (been a long time since I've used this library...). I discovered that the problem was that I didn't specify a valid config for the `swagger` part of the configuration, even though I was trying to generate routes for the test.

The obvious solution is to put a dummy (but valid) configuration in both sections, but this just feels a little backwards to me. If you're not using a feature, why should you be required to configure it?

**Comment by WoH** [comment]
I can confirm the issue and agree with the assessment. In my case, the swagger docs are created dynamically, so I don't load the swagger config from the tsoa file. I'm using dummy settings there but ideally I would not have to do that.

**Comment by dgreene1** [comment]
This admittedly strange result was the only possible fix for https://github.com/lukeautry/tsoa/issues/420

The shortest summary I can give is:
1) dynamic controller globs was added for generateRoutes. This required the new property to live on RouteConfig
2) immediately after releasing that, the developer of the dynamic route generation feature pointed out to me that CLI users that used this feature would cause the swagger.yaml to be empty since the glob ily worked on the route side.
3) the fix was to allow the dynamic glob property to be added for swagger generation too
4) Here’s the rub... what if the dynamic route glob property was different for swagger and for the routes and both methods were called in the same node process (see the tsoa internal cache PR for more details on why this would be terrible). So the solution to that was to require the full config to be submitted for both generate methods so that the dynamic glob property could be validated.

That’s why the second parameter of RouteConfig was added to generateSwagger (or whatever that method is called (I’m not at my desk now).

I recognize that this is not a pleasing design choice.

But my question for you both is: what is your recommendation?

**Comment by dgreene1** [comment]
I believe the best solution would be to create a new method that has this signature:
```ts
function generateSwaggerAndRoutes(config: Config)
```

This would future proof things so that we can introduce any change to the inputs by simply adding a new property to the Config interface. Please advise if this would eliminate the confusion you’re having @HoldYourWaffle.

**Comment by dgreene1** [comment]
> If you're not using a feature, why should you be required to configure it?

The noImplicitAdditionalProperties is something that impacts both swagger generation and the validation that is built dynamically when the routes are generated.

That’s one use case where configs from one of the two commands impacts the other command. Another use case is the one I described above where the dynamic generation glob started on the routes config and then we realized that it was needed for both generation.

Long story short, unless this can be clarified as an actual issue/bug, I feel that this is something we might need to close as “not a defect.”

I realize that it has poor DX (developer user experience), however the config data is required for some use cases.

That being said, if someone can recommend a way that the two use cases can still function and we can simultaneously improve the DX, then I’m happy to accept the suggestion.

**Comment by dgreene1** [comment]
Just follow up @HoldYourWaffle / @WoH so is there something actionable here?

**Comment by HoldYourWaffle** [comment]
Wouldn't it be easier to move the `controllerPathGlobs` and `noImplicitAdditionalProperties` options "one level up" like this:
```jsonc
{
	// global properties
	"noImplicitAdditionalProperties": true,
	"controllerPathGlobs": [
		"./controllers/public/**/*",
		"./controllers/private/**/*"
	],
	
	"swagger": {
		// properties for swagger generation
		"outputDirectory": "./my-amazing-swagger-spec"
	},
	
	"routes": {
		// properties for route generation
		"routesDir": "./my-amazing-routes"
	}
}
```
This might require a small refactor of the "configuration system", but the last time I checked (which was admittedly not at my brightest moment) it looked like that would be a welcome improvement anyway. This might be what you're trying to say with but I couldn't tell:

> I believe the best solution would be to create a new method that has this signature:
> 
> ```ts
> function generateSwaggerAndRoutes(config: Config)
> ```
> 
> This would future proof things so that we can introduce any change to the inputs by simply adding a new property to the Config interface.

This sounds to me like a solution for a different problem, but that might just be my morning head.

**Comment by dgreene1** [comment]
@HoldYourWaffle. What was the precise error message that you received?

> While testing for #192 I struggled a bit with getting the configuration right (been a long time since I've used this library...). I discovered that the problem was that I didn't specify a valid config for the swagger part of the configuration

Because if you’re not getting an error, then I can’t say that this is really an issue and therefore it will get closed.

**Comment by HoldYourWaffle** [comment]
I don't remember exactly (not at my desk right now so I can't test it) but they were all in the form of:
```
Can't access property 'x' from undefined
```

Is there a reason why my "move up a level" suggestion wouldn't work? Even if it wouldn't fit your definition of an issue it would still be a welcome enhancement in my opinion.

I also read your comment again to see if there was anything I missed
> So the solution to that was to require the full config to be submitted for both generate methods so that the dynamic glob property could be validated.

I'm not sure what you mean by that, what validation has to be done?

**Comment by WoH** [comment]
I'd argue even if it's not a bug, it is still an issue that should stay open since it is confusing behavior.

These are the errors if you're trying to generate routes with only route specific config.

```
Generate routes error.
 TypeError: Cannot read property 'outputDirectory' of undefined
    at Object.<anonymous> ([redacted]/node_modules/tsoa/dist/cli.js:123:29)
    at step ([redacted]/node_modules/tsoa/dist/cli.js:33:23)
    at Object.next ([redacted]/node_modules/tsoa/dist/cli.js:14:53)
    at [redacted]/node_modules/tsoa/dist/cli.js:8:71
    at new Promise (<anonymous>)
    at __awaiter ([redacted]/node_modules/tsoa/dist/cli.js:4:12)
    at Object.exports.validateSwaggerConfig ([redacted]/node_modules/tsoa/dist/cli.js:118:60)
    at Object.<anonymous> ([redacted]/node_modules/tsoa/dist/cli.js:339:50)
    at step ([redacted]/node_modules/tsoa/dist/cli.js:33:23)
    at Object.next ([redacted]/node_modules/tsoa/dist/cli.js:14:53)
```

and 

```
Generate routes error.
 Error: Missing entryFile: Configuration must contain an entry point file.
    at Object.<anonymous> ([redacted]/node_modules/tsoa/dist/cli.js:127:27)
    at step ([redacted]/node_modules/tsoa/dist/cli.js:33:23)
    at Object.next ([redacted]/node_modules/tsoa/dist/cli.js:14:53)
    at [redacted]/node_modules/tsoa/dist/cli.js:8:71
    at new Promise (<anonymous>)
    at __awaiter ([redacted]/node_modules/tsoa/dist/cli.js:4:12)
    at Object.exports.validateSwaggerConfig ([redacted]/node_modules/tsoa/dist/cli.js:118:60)
    at Object.<anonymous> ([redacted]/node_modules/tsoa/dist/cli.js:339:50)
    at step ([redacted]/node_modules/tsoa/dist/cli.js:33:23)
    at Object.next ([redacted]/node_modules/tsoa/dist/cli.js:14:53)
```

**Comment by lukeautry** [comment]
I agree with pulling up the relevant config properties into the root config object. Swagger generation should never require route generation config and vice versa. Unfortunately, fixing this will be a breaking change, but the features we're talking about are so new that it should be relatively easy for consumers to get updated.

**Comment by WoH** [comment]
@lukeautry Maybe it makes sense to collect all the (mostly small) breaking changes, work on a milestone and release them as 3.0?

**Comment by lukeautry** [comment]
Yes, I think so.

**Comment by HoldYourWaffle** [comment]
I think a new major version release to enable a good cleanup of these kinds of small annoyances would definitely be a good thing.

**Comment by dgreene1** [comment]
## Proposed short-term solution:

Until version 3.0 is released, would it alleviate your concerns if we:
* made the swagger config optional on `generateRoutes`?
* and made the route config optional on `generateSwaggerSpec`?

That way you don't have to pass in `{}` and you can pass in `null` to be more explicit?

**Comment by HoldYourWaffle** [comment]
That looks like a valid short term solution to me.

**Comment by dgreene1** [comment]
We’ll fix this in v3
